src/org/zalando/stups/friboo/system/new_http.clj

```diff
@@ -0,0 +1,275 @@
+  (:require [io.sarnowski.swagger1st.executor :as s1stexec]
+            [io.clj.logging :refer [with-logging-context]]
+            [ring.util.response :as r]
+            [org.zalando.stups.friboo.ring :as ring]
+            [clojure.data.codec.base64 :as b64]
+            [org.zalando.stups.friboo.log :as log]
+            [io.sarnowski.swagger1st.util.api :as api]
+            [io.sarnowski.swagger1st.core :as s1st]
+            [io.sarnowski.swagger1st.util.api :as s1stapi]
+            [ring.adapter.jetty :as jetty]
+            [org.zalando.stups.friboo.security :as security]
+            [ring.middleware.gzip :as gzip])
+  (:import (com.stuartsierra.component Lifecycle)
+           (com.netflix.hystrix.exception HystrixRuntimeException)
+           (org.zalando.stups.txdemarcator Transactions)))
+
+(defn merged-parameters
+  "According to the swagger spec, parameter names are only unique with their type. This one assumes that parameter names
+   are unique in general and flattens them for easier access."
+  [request]
+  (apply merge (vals (:parameters request))))
+
+(defn make-resolver-fn [controller]
+  "Calls operationId function with controller, flattened request params and raw request map."
+  (fn [request-definition]
+    (when-let [operation-fn (s1stexec/operationId-to-function request-definition)]
+      (fn [request]
+        (operation-fn controller (merged-parameters request) request)))))
+
+(defn middleware-chain [context middlewares]
+  (reduce #(%2 %1) context middlewares))
+
+(defn flatten1
+  "Flattens the collection one level, for example, converts {:a 1 :b 2} to (:a 1 :b 2)."
+  [coll]
+  (apply concat coll))
+
+(defn with-flattened-options-map
+  "If f is defined like this:
+
+  `(defn f [first-arg & {:keys [a b] :as opts}] ... )`
+
+  makes it convenient to pass `opts` as a map:
+
+  `(-> first-arg (with-flattened-options-map f {:a 1 :b 2}))`
+  will result in
+  `(f first-arg :a 1 :b 2)`
+
+  Used in `start-component` for clarity.
+  "
+  [first-arg f options]
+  (apply f first-arg (flatten1 options)))
+
+(defn start-component [{:as this :keys [api-resource configuration middlewares security-handlers controller s1st-options]}]
+  (log/info "Starting HTTP daemon for API %s" api-resource)
+  (when (:handler this)
+    (throw (ex-info "Component already started, aborting." {})))
+  (let [handler (-> (apply s1st/context :yaml-cp api-resource (flatten1 {:context s1st-options}))
+                    (assoc :component this)
+
+                    (middleware-chain (:before-discoverer middlewares))
+                    (with-flattened-options-map s1st/discoverer (:discoverer s1st-options))
+
+                    (middleware-chain (:before-mapper middlewares))
+                    (s1st/mapper)
+
+                    (middleware-chain (:before-parser middlewares))
+                    (with-flattened-options-map s1st/parser (:parser s1st-options))
+
+                    (middleware-chain (:before-protector middlewares))
```

Comment:
We should really check if the protection step can already happen before the parsing. I don't see why we would need parsed parameters in order to evaluate the Authorization header